src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionXmlFormatReader.cs

```diff
@@ -152,15 +152,27 @@ private void ReflectionSetMemberValue(object obj, object memberValue, MemberInfo
                         Type genericCollectionType = Globals.TypeOfICollectionGeneric.MakeGenericType(itemType);
                         if (genericCollectionType.IsAssignableFrom(propertyType))
                         {
-                             addMethod = genericCollectionType.GetMethod("Add");
+                            addMethod = genericCollectionType.GetMethod("Add");
+                        }
+                        else if (Globals.TypeOfIDictionary.IsAssignableFrom(propertyType) && Globals.TypeOfIDictionary.IsAssignableFrom(memberValue.GetType()))
+                        {
+                            var sourceDict = (IDictionary)memberValue;
+                            var targetDict = (IDictionary)property;
+                            foreach (var key in sourceDict.Keys)
+                            {
+                                var value = sourceDict[key];
+                                targetDict.Add(key, value);
```

Comment:
Dictionary.Add can throw if the item is already there. Can sourceDict have duplicate keys? There are implementations of dictionaries which allow duplicate keys, could they end up being used here? What happens if there's a duplicate key in the source XML?